scipy/sparse/linalg/_eigen/_svds.py

```diff
@@ -327,13 +337,45 @@ def matmat_XH_X(x):
     u = u[:, ::-1]
     s = s[::-1]
     vh = vh[::-1]
+
     jobu = return_singular_vectors in {True, 'u'}
     jobv = return_singular_vectors in {True, 'vh'}
-    if not transpose:
-        u = u if jobu else None
-        vh = vh @ _herm(eigvec) if jobv else None
+
+    if tol is None or tol <= 0.0:
+        res_tol = np.sqrt(1e-15) * max(A.shape)
     else:
+        res_tol = tol
+
+    if transpose:
         u_tmp = eigvec @ _herm(vh) if jobu else None
-        vh = _herm(u) if jobv else None
+        if jobv:
+            # if singular values are very small, the right singular vectors
+            # computed above may be inaccurate, so we have to re-compute
+            # them as the eigenvectors of X_XH
+            eigvals, eigvec = lobpcg(X_XH, u, tol=res_tol**2, maxiter=maxiter,
```

Comment:
@evgueni-ovtchinnikov why squaring the res_tol here and below?

Reply:
@lobpcg just for consistency with the other eigensolver calls

I actually objected against squaring, if you remember, but agreed to make a separate issue for the tolerance.